File: cpp/0104.0_Maximum_Depth_of_Binary_Tree.cpp
/*
DFS

Runtime: 3 ms, faster than 96.69% of C++ online submissions for Maximum Depth of Binary Tree.
Memory Usage: 18.9 MB, less than 15.07% of C++ online submissions for Maximum Depth of Binary Tree.
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (nullptr == root) {
            return 0;
        }
        return 1 + max(maxDepth(root->left), maxDepth(root->right));
    }
};




/*
DFS

Runtime: 7 ms, faster than 83.32% of C++ online submissions for Maximum Depth of Binary Tree.
Memory Usage: 18.7 MB, less than 89.47% of C++ online submissions for Maximum Depth of Binary Tree.
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        return root ? 1 + max(maxDepth(root->left), maxDepth(root->right)) : 0;
    }
};

File: csharp/0104.0_Maximum_Depth_of_Binary_Tree.cs
/*
Runtime: 86 ms, faster than 89.37% of C# online submissions for Maximum Depth of Binary Tree.
Memory Usage: 39.6 MB, less than 62.60% of C# online submissions for Maximum Depth of Binary Tree.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int MaxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int l = MaxDepth(root.left);
        int r = MaxDepth(root.right);
        return 1 + Math.Max(l, r);
    }
}



/*
Runtime: 83 ms, faster than 95.23% of C# online submissions for Maximum Depth of Binary Tree.
Memory Usage: 39.5 MB, less than 62.60% of C# online submissions for Maximum Depth of Binary Tree.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int MaxDepth(TreeNode root) {
        return root == null ? 0 : 1 + Math.Max(MaxDepth(root.left), MaxDepth(root.right));
    }
}